overlord/configstate/config/helpers.go

```diff
@@ -116,3 +119,107 @@ func GetFromChange(snapName string, subkeys []string, pos int, config map[string
+
+	// Get current configuration of the snap from state
+	err := st.Get("config", &config)
+	if err == state.ErrNoState {
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("internal error: cannot unmarshal configuration: %v", err)
+	}
+	if snapcfg, ok := config[snapName]; ok {
+		err = st.Get("config-snapshots", &configSnapshots)
+		if err == state.ErrNoState {
+			configSnapshots = make(map[string]map[string]interface{})
+		} else if err != nil {
+			return err
+		}
+		cfgs := configSnapshots[snapName]
+		if cfgs == nil {
+			cfgs = make(map[string]interface{})
+		}
+		cfgs[rev.String()] = snapcfg
+		configSnapshots[snapName] = cfgs
+		st.Set("config-snapshots", configSnapshots)
+	} // else - no config, nothing to do
+	return nil
+}
+
+// RestoreConfigurationSnapshotMaybe restores a given revision of snap configuration into config -> snapName.
+// If no configuration exists for given revision it does nothing (no error).
+func RestoreConfigurationSnapshotMaybe(st *state.State, snapName string, rev snap.Revision) error {
+	var config map[string]interface{}                     // snap => configuration
+	var configSnapshots map[string]map[string]interface{} // snap => revision => configuration
+
+	err := st.Get("config-snapshots", &configSnapshots)
+	if err == state.ErrNoState {
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("internal error: cannot unmarshal config-snapshots: %v", err)
+	}
+
+	err = st.Get("config", &config)
+	if err == state.ErrNoState {
+		config = make(map[string]interface{})
+	} else if err != nil {
+		return fmt.Errorf("internal error: cannot unmarshal configuration: %v", err)
+	}
+
+	if cfg, ok := configSnapshots[snapName]; ok {
+		if revCfg, ok := cfg[rev.String()]; ok {
+			config[snapName] = revCfg
+			st.Set("config", config)
+		}
+	}
+
+	return nil
+}
+
+// DeleteConfigurationSnapshotMaybe removes configuration snapshot of given snap/revision.
+// If no configuration exists for given revision it does nothing (no error).
+func DeleteConfigurationSnapshotMaybe(st *state.State, snapName string, rev snap.Revision) error {
+	var configSnapshots map[string]map[string]interface{} // snap => revision => configuration
+	err := st.Get("config-snapshots", &configSnapshots)
+	if err == state.ErrNoState {
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("internal error: cannot unmarshal config-snapshots: %v", err)
+	}
+
+	if revCfgs, ok := configSnapshots[snapName]; ok {
+		delete(revCfgs, rev.String())
+		if len(revCfgs) == 0 {
+			delete(configSnapshots, snapName)
+		} else {
+			configSnapshots[snapName] = revCfgs
+		}
+		st.Set("config-snapshots", configSnapshots)
+	}
+	return nil
+}
+
+// DeleteSnapConfig removed configuration of given snap from the state.
+func DeleteSnapConfig(st *state.State, snapName string) error {
+	var config map[string]map[string]*json.RawMessage // snap => key => value
+
+	st.Lock()
+	defer st.Unlock()
```

Comment:
this is strange, usually the caller would care for locking, it seems doDiscardSnap has grown quite a bit and its strategy with dealing with locks doesn't make much sense anymore, it should probably unlock only around removing files